### ISSUE TITLE
Accessibility checker dialog cannot be closed by clicking the X button

### ISSUE THREAD

**Issue by laymonage** [wagtail/wagtail#10924]
### Issue Summary

After opening the accessibility checker dialog in the Wagtail userbar, clicking the X button does not close the dialog (it does nothing). Clicking the userbar menu item again does close it, though.

### Steps to Reproduce

1. Spin up bakerydemo
2. Log in to the Wagtail admin
3. Open the bakerydemo home page
4. Open the Wagtail userbar
5. Click on the "Acessibility" item
6. Try to click on the X button to close the dialog
7. Observe that it does not close the dialog

I haven't confirmed it but this might be a regression in #10397.

**Comment by lb-** [comment]
Hey @laymonage - I just tried in Firefox 115, Chrome 116, Safari 16.1 and iOS (Simulator only) - I could not find an issue with the close button.

Maybe it's a specific browser or maybe something odd with the production JS build compared to development? 

Do you recall what browser it was that the issue was in.

**Comment by laymonage** [comment]
Hi @lb-, thanks and sorry I forgot to specify the browsers. Tested on Chrome 117, Firefox 117, and Safari 16.6 on macOS Ventura 13.5.2, all have the same issue. I tried building the static assets with `npm start` and `npm run build`, same issue. Tried deleting `node_modules` and the built `static` folder then reinstalling+rebuilding the assets, still the same.

Not sure if it's something up with my setup...

I'm also running on Django 4.2.5, if that helps.

**Comment by lb-** [comment]
Very odd, and the close buttons within Wagtail admin all work fine?

**Comment by thibaudcolas** [comment]
I can reproduce the issue as well, in `main` but not in v5.1.1. From what I can see it’s a regression from [`a11y-dialog` v8.0.0](https://github.com/KittyGiraudel/a11y-dialog/releases/tag/8.0.0) via #10748, specifically this:

```
Click events are now listened to at the document level (by @mxmason in https://github.com/KittyGiraudel/a11y-dialog/pull/387, solving https://github.com/KittyGiraudel/a11y-dialog/issues/367).
```

Inspecting the 5.1.1 and main implementations with a debugger, when the click event triggers a11y-dialog’s listener, the click event’s target is reported as our `wagtail-userbar` web component rather than the button – which then doesn’t work with a11y-dialog’s logic.

---

I’m not super familiar with event delegation / bubbling through the shadow DOM / custom elements but here are the options I can envision to fix this:

1. If this is a bug in `a11y-dialog`, get them to switch from [`event.target`](https://github.com/KittyGiraudel/a11y-dialog/blob/ea5d251c1b814c873fe024df0dde8fce9e01de98/src/a11y-dialog.ts#L167) to something like `event.composedPath()[0]` in their click event handling
2. In the meantime / if it’s more complex than it seems, add our own click event handling which uses a11y-dialog’s `.hide()` API.
3. Some kind of configuration change to how our custom element is declared that retains the styles encapsulation as-is but allows more DOM APIs to "pass through". We already use Shadow DOM in "open" mode, but I see there are other options to [`attachShadow`](https://developer.mozilla.org/en-US/docs/Web/API/Element/attachShadow#parameters) and perhaps other things outside that I’m not aware of.

See also #10674 which is another issue specific to our code being inside a web component.

**Comment by lb-** [comment]
Great find. Very confused why I can't reproduce it manually though but that makes sense. 

Should we just revert a11y-dialog?

Alternatively, adding `data-action="w-dialog#hide"` to the button should probably fix this. It'll use Stimulus' event listener and attach to the close button DOM element.

**Comment by lb-** [comment]
Note - Thibaud has raised https://github.com/KittyGiraudel/a11y-dialog/issues/582

I have a PR coming with the fix recommended 
> In the meantime / if it’s more complex than it seems, add our own click event handling which uses a11y-dialog’s .hide() API.

I have been able to reproduce this, I have had some issues with my Docker node_modules caching it seems, will try to work out why and update the Docker repo with better instructions.

**Comment by digicase** [comment]
Possibly related to to this issue, I experienced a small bug. When using Safari 17 on MacOS, if I open the accessibility checker, and then try to open the Wagtail user bar menu, the menu will animate open but then disappear before it can be used. Closing the accessibility checker enables opening the user bar menu as expected. I am using the bakery demo in Docker, Wagtail v5.1.2
